tests/beaker_tests/lib/utilitylib.rb

```diff
@@ -455,87 +456,159 @@ def bgp_nbr_remote_as(agent, remote_as)
+#   parameter keys in the manifest. When these are used the puppet resource
+#   command string becomes a combination of the title pattern and these params.
+#
+def create_manifest_and_resource(tests, id)
+  fail 'tests[:resource_name] is not defined' unless tests[:resource_name]
+  tests[id][:title_pattern] = id if tests[id][:title_pattern].nil?
 
-  t[:afi], t[:vrf] = title.split
-  t.merge!(tests[id][:pim])
-  t[:vrf] = 'default' if t[:vrf].nil?
-  t
-end
+  # Create the cmd string for puppet_resource
+  puppet_resource_cmd_from_params(tests, id)
 
-# If a [:title] exists merge it with the [:pim_rp] values to create a complete pim.
-def pim_rp_title_pattern_munge(tests, id)
-  title = tests[id][:title_pattern]
-  pim_rp = tests[id][:pim_rp]
+  # Create any title-params manifest entries. Typically only used
+  # for title-pattern testing
+  manifest = prop_hash_to_manifest(tests[id][:title_params])
 
-  if title.nil?
-    puts 'no title'
-    return pim_rp
-  end
+  # Setup the ensure state, manifest string, and resource command state
+  if tests[id][:ensure] == :absent
+    state = 'ensure => absent,'
+    tests[id][:resource] = { 'ensure' => 'absent' }
+  else
+    state = 'ensure => present,'
+    # Create the property string for the manifest
+    manifest += prop_hash_to_manifest(tests[id][:manifest_props]) if
+      tests[id][:manifest_props]
 
-  tests[id][:pim_rp] = {} if pim_rp.nil?
-  t = {}
+    # Automatically create a hash of expected states for puppet resource
+    # -or- use a static hash
+    tests[id][:resource] = tests[id][:manifest_props] unless
+      tests[id][:resource]
+  end
 
-  t[:afi], t[:vrf], t[:rp_addr] = title.split
-  t.merge!(tests[id][:pim_rp])
-  t[:vrf] = 'default' if t[:vrf].nil?
-  t
+  tests[id][:manifest] = "cat <<EOF >#{PUPPETMASTER_MANIFESTPATH}
+  \nnode default {
+  #{tests[:resource_name]} { '#{tests[id][:title_pattern]}':
+    #{state}\n#{manifest}
+  }\n}\nEOF"
 end
 
-# If a [:title] exists merge it with the [:pim_group] values to create a complete pim.
-def pim_group_title_pattern_munge(tests, id)
-  title = tests[id][:title_pattern]
-  pim_group = tests[id][:pim_group]
+# test_harness_run
+#
+# This method is a front-end for test_harness_common & create_manifest_and_resource,
+# as well as testbed cleanup.
+def test_harness_run(tests, id)
+  return unless platform_supports_test(tests, id)
 
-  if title.nil?
-    puts 'no title'
-    return pim_group
-  end
+  tests[id][:ensure] = :present if tests[id][:ensure].nil?
 
-  tests[id][:pim_group] = {} if pim_group.nil?
-  t = {}
+  # Build the manifest for this test
+  create_manifest_and_resource(tests, id)
 
-  t[:afi], t[:vrf], t[:rp_addr], t[:group] = title.split
-  t.merge!(tests[id][:pim_group])
-  t[:vrf] = 'default' if t[:vrf].nil?
-  t
+  resource_absent_cleanup(agent, tests[id][:preclean]) if
+    tests[id][:preclean]
+
+  # Set up remote-as if necessary
+  bgp_nbr_remote_as(agent, tests[id][:remote_as]) if tests[id][:remote_as]
```

Comment:
I agree. These dependencies are bound to grow so better to move them into their own method.